plugins/zmq/src/plugin.cpp

```diff
@@ -0,0 +1,415 @@
+    }
+
+  private:
+    monitor(::zmq::socket_t&& socket) : monitor_socket_{std::move(socket)} {
+    }
+
+    ::zmq::socket_t monitor_socket_;
+  };
+
+public:
+  static auto make_source(const connector_args& args) -> caf::expected<engine> {
+    try {
+      auto result = engine{::zmq::socket_type::sub};
+      if (args.bind)
+        result.bind(args.endpoint);
+      else
+        result.connect(args.endpoint);
+      const auto* everything = "";
+      result.socket_.set(::zmq::sockopt::subscribe, everything);
+      return result;
+    } catch (const ::zmq::error_t& e) {
+      return make_error(e);
+    }
+  }
+
+  static auto make_sink(const connector_args& args) -> caf::expected<engine> {
+    try {
+      auto result = engine{::zmq::socket_type::pub};
+      if (args.connect)
+        result.connect(args.endpoint);
+      else
+        result.bind(args.endpoint);
+      return result;
+    } catch (const ::zmq::error_t& e) {
+      return make_error(e);
+    }
+  }
+
+  auto send(chunk_ptr chunk) -> caf::error {
+    try {
+      TENZIR_DEBUG("waiting until socket is ready to send");
+      if (not poll(socket_, ZMQ_POLLOUT))
+        return caf::make_error(ec::timeout, "timed out while polling socket");
+      auto message = ::zmq::message_t{*chunk};
+      auto flags = ::zmq::send_flags::none;
+      auto bytes = socket_.send(message, flags);
+      TENZIR_ASSERT(bytes); // only nullopt in non-blocking mode.
+      TENZIR_DEBUG("sent message with {} bytes", *bytes);
+      return {};
+    } catch (const ::zmq::error_t& e) {
+      return make_error(e);
+    }
+  }
+
+  auto receive() -> caf::expected<chunk_ptr> {
+    try {
+      TENZIR_DEBUG("waiting until socket is ready to receive");
+      if (not poll(socket_, ZMQ_POLLIN))
+        return caf::make_error(ec::timeout, "timed out while polling socket");
+      auto message = std::make_shared<::zmq::message_t>();
+      auto flags = ::zmq::recv_flags::none;
+      auto bytes = socket_.recv(*message, flags);
+      TENZIR_ASSERT(bytes); // only nullopt in non-blocking mode.
+      TENZIR_DEBUG("got 0mq message with {} bytes", *bytes);
+      const auto* data = message->data();
+      auto size = message->size();
+      auto deleter = [msg = std::move(message)]() noexcept {};
+      return chunk::make(data, size, deleter);
+    } catch (const ::zmq::error_t& e) {
+      return make_error(e);
+    }
+  }
+
+private:
+  static auto make_error(const ::zmq::error_t& error) -> caf::error {
+    return caf::make_error(ec::unspecified,
+                           fmt::format("ZeroMQ: {} ({})", error.what(),
+                                       error.num()));
+  }
+
+  static auto poll(::zmq::socket_t& socket, short flags,
+                   std::optional<std::chrono::milliseconds> timeout = {})
+    -> bool {
+    auto items = std::array<::zmq::pollitem_t, 1>{
+      {{socket.handle(), 0, flags, 0}},
+    };
+    auto ms = timeout ? *timeout : std::chrono::milliseconds(-1);
```

Comment:
Can you add a comment here that explains what -1ms indicates?

Reply:
45e726759e